Backports for 3.4.3

mattpap opened the pull request:
This PR collects bug-fixes and tasks for a 3.4.3 release:

- [x] https://github.com/bokeh/bokeh/pull/13952
- [x] https://github.com/bokeh/bokeh/pull/13949
- [x] https://github.com/bokeh/bokeh/pull/13970
- [x] https://github.com/bokeh/bokeh/pull/13974
- [x] https://github.com/bokeh/bokeh/pull/13975
- [x] https://github.com/bokeh/bokeh/pull/13914
- [x] https://github.com/bokeh/bokeh/pull/13978
- [x] https://github.com/bokeh/bokeh/pull/13966
- [x] https://github.com/bokeh/bokeh/pull/13985
- [x] https://github.com/bokeh/bokeh/pull/13989
- [x] 3.4.3 release notes
- [x] updated switcher.json

Release checklist:

- [x] do **not** squash merge
- [x] update milestone to `3.4.3` of relevant issues and PRs
- [x] remove `NEEDS BACKPORT` label

mosc9575 committed:
update type hints in gridplot for children (#13914)

* update type hints

* add overload

* add test_layouts__typing.py

* fix mypy no-untyped-def error

* add @overload to exclude_also for coverage report

mosc9575 committed:
add release to project configuration (#13949)

* add release to project configuration

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

---------

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

mattpap committed:
Implement data inheritance in `ImageBase._set_data()` (#13952)

* Implement data inheritance in ImageBase._set_data()

* Rename {_width,_height} -> image_{width,height}

* Add unit regression tests

mosc9575 commented:
We should add

- [x] https://github.com/bokeh/bokeh/pull/13949

to the  3.4.3 release. Otherwise the version banner will warn about **an unknown version** again.

mattpap committed:
Update switcher.json

mattpap committed:
Robustify static case in `may_have_refs()` (#13970)

* Robustify static case in may_have_refs()

* Add cross regression tests

mattpap committed:
Update location of *.d.ts files in package.json (#13975)

mattpap committed:
Restore support for BOKEH_MINIFIED=no in Resources (#13974)

mattpap committed:
Always proxy tools (even if only one) when merging toolbars (#13978)

mattpap committed:
Make sure integer indices are used when indexing categories (#13966)

* Make sure integer indices are used when indexing categories

* Add unit regression tests

s-t-e-v-e-n-k committed:
Use non-deprecated argument in pytest hook (#13985)

pytest_report_collectionfinish()'s startdir argument has been deprecated
since 7.0, and with pytest 8.2 now raises a deprecation warning that
it will be removed in 9.0. Switch to the new start_path argument.

bryevdv committed:
fixes for mypy 1.11 (#13989)

mattpap committed:
Add release notes

mattpap committed:
numpy 2.0 compatibility
